docs/source/services.md

```diff
@@ -0,0 +1,84 @@
+# JupyterHub services
+
+JupyterHub 0.7 adds the notion of Services.
+A Service is a process that interacts with the Hub REST API.
+Services may perform actions such as shutting down user servers that have been idle for some time,
+or registering additional web servers that should also use the Hub's authentication
+and be served behind the Hub's proxy.
+
+There are two main characteristics of services:
+
+1. Is it **managed** by JupyterHub?
+2. Does it have a web server that should be added to the proxy?
+
+If a `command` is specified for launching the service, it will be started and managed by the Hub.
+If a `url` is specified for where the service runs its own webserver,
+it will be added to the Hub's proxy at `/service/:service-name`.
+
+## Managed services
+
+**Managed** services are services that the Hub starts and is responsible for.
+These can only be local subprocesses of the Hub.
+While there are similarities with notebook Spawners,
+there are no plans to support the same spawning abstractions as notebook
+If you want to run these services in docker or other environments,
```

Comment:
For generic setup and teardown hooks, do you think there's a benefit of having configurable extension points vs telling advanced users to subclass and override `JupyterHub.initialize` and `JupyterHub.cleanup`?  Flask, for example, [explicitly encourages](http://flask.pocoo.org/docs/0.11/becomingbig/#becoming-big) advanced users to override documented APIs on their main application class.  

I think the main downside to a subclass/extend strategy for us would be that you have to write your own entrypoint to use a subclass, since the default `jupyterhub` cli calls `JupyterHub.launch_instance` directly. (I'm running from my own entrypoint on Quantopian, which is easy enough if you know where to look.)

I'd be :+1: on supporting passing instances of `Service` to set up services that don't fall cleanly into the existing buckets of managed/non-managed.  If that were the API though, seems a little odd to also support passing dictionaries, since those are presumably getting turned into `Service` objects by the hub anyway.

What would you think about making `Service` an interface (consisting, say, of `start`, `stop`, and `close`), and defining two concrete implementations `LocalProcessService` and `ExternalService`, which would express the two currently-targeted cases, but would leave room for enterprising third-party developers to write, say, `LocalDockerService` or `PeriodicalCallbackService`?  This would also make it easier to write services as jupyterhub plugins, since you could just tell users to put something like the following in their jupyterhub config:
`
```
from my_cool_module import MyCoolService

c = get_config()
c.JupyterHub.services = MyCoolService(...)
```